File: tests/argusflow-windows/support/observation_demo/conversation/tools.mjs
/** 模型可调用的白名单查询，只能补读现有证据，不能操作桌面或网络。 */
import { inspectFacts } from "./evidence.mjs";

export const evidenceTools = [
  {
    type: "function",
    function: {
      name: "inspect_evidence",
      description:
        "按时间线 ID 查询原始 UIA 文档/选区、剪贴板或 CDP 事实。请求最多24个ID，每页返回6条和remaining_ids；有剩余时可继续查询。",
      parameters: {
        type: "object",
        properties: {
          ids: {
            type: "array",
            items: { type: "string" },
            minItems: 1,
            maxItems: 24,
          },
        },
        required: ["ids"],
        additionalProperties: false,
      },
    },
  },
  {
    type: "function",
    function: {
      name: "view_change",
      description:
        "查看 sample ID 与同窗口前一帧之间的变化区域前后小图。context=change 仅差分；看不全时可用 text_line 包含相交文字行。禁止整屏，最多三对，单图最大640px宽及180k像素。",
      parameters: {
        type: "object",
        properties: {
          sample_id: { type: "string" },
          context: { type: "string", enum: ["change", "text_line"] },
        },
        required: ["sample_id"],
        additionalProperties: false,
      },
    },
  },
];

export async function dispatchTool(recording, images, call) {
  if (
    typeof call.function?.arguments !== "string" ||
    call.function.arguments.length > 2048
  )
    throw Error("工具参数无效或超限");
  const args = JSON.parse(call.function.arguments);
  switch (call.function.name) {
    case "inspect_evidence":
      if (Object.keys(args).some((k) => k !== "ids"))
        throw Error("未知工具参数");
      if (!Array.isArray(args.ids) || args.ids.length > 24)
        throw Error("查询参数超出上限");
      return {
        facts: inspectFacts(recording, args.ids.slice(0, 6)),
        remaining_ids: args.ids.slice(6),
        images: [],
      };
    case "view_change":
      if (Object.keys(args).some((k) => !["sample_id", "context"].includes(k)))
        throw Error("未知工具参数");
      return images.change(args.sample_id, args.context);
    default:
      throw Error("工具不在只读白名单");
  }
}

File: crates/argusflow-ai/src/inference/tools.rs
//! 模型工具只读冻结证据；不接受任意路径、脚本或网络地址。
use crate::{AiError, Evidence, Result};
use serde::Deserialize;
use serde_json::{Value, json};
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Ids {
    ids: Vec<String>,
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Frame {
    id: String,
}
pub(crate) fn definitions(vision: bool) -> Vec<Value> {
    let mut tools = vec![
        json!({"type":"function","function":{"name":"inspect_evidence","description":"根据时间线ID读取1到6条完整事实","parameters":{"type":"object","properties":{"ids":{"type":"array","items":{"type":"string"},"minItems":1,"maxItems":6}},"required":["ids"],"additionalProperties":false}}}),
    ];
    if vision {
        tools.push(json!({"type":"function","function":{"name":"view_change","description":"视觉记录ID的同来源前后局部变化图，最多三对","parameters":{"type":"object","properties":{"id":{"type":"string"}},"required":["id"],"additionalProperties":false}}}));
    }
    tools
}
pub(crate) fn dispatch(
    evidence: &Evidence,
    call: &Value,
    vision: bool,
    pairs: &mut u8,
) -> Result<(Value, Vec<Value>, u64)> {
    let name = call["function"]["name"]
        .as_str()
        .ok_or_else(|| AiError::Invalid("工具名称无效".into()))?;
    let arguments = call["function"]["arguments"]
        .as_str()
        .filter(|s| s.len() <= 2048)
        .ok_or_else(|| AiError::Invalid("工具参数超限".into()))?;
    match name {
        "inspect_evidence" => Ok((
            evidence.inspect(&serde_json::from_str::<Ids>(arguments)?.ids)?,
            vec![],
            0,
        )),
        "view_change" if vision && *pairs < 3 => {
            *pairs += 1;
            let (images, pixels) =
                evidence.change(&serde_json::from_str::<Frame>(arguments)?.id)?;
            Ok((
                json!({"images":"随后 user 消息中的前后图片属于此次调用"}),
                images,
                pixels,
            ))
        }
        _ => Err(AiError::Invalid("工具不在允许范围内或图片预算耗尽".into())),
    }
}
